server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/DependencyGraph.java

```diff
@@ -26,36 +28,36 @@
 import org.jgrapht.graph.DefaultDirectedGraph;
 import org.jgrapht.graph.DefaultEdge;
 import org.jgrapht.graph.builder.DirectedGraphBuilder;
-import org.jgrapht.traverse.AbstractGraphIterator;
 import org.jgrapht.traverse.TopologicalOrderIterator;
 
 public class DependencyGraph<T> {
 
-
     private final DirectedGraphBuilder<T, DefaultEdge, DefaultDirectedGraph<T, DefaultEdge>> builder;
-
-    private final Function<T, String> getName;
-    private final Function<T, T> getParent;
+    private final Function<T, Optional<T>> getParent;
 
 
-    public DependencyGraph(Function<T, String> getName, Function<T, T> getParent) {
-        this.getName = getName;
+    public DependencyGraph(Function<T, Optional<T>> getParent) {
         this.getParent = getParent;
         this.builder = new DirectedGraphBuilder<>(new DefaultDirectedGraph<>(DefaultEdge.class));
     }
 
     public void registerItem(T item) {
         builder.addVertex(item);
-        T parentNode = getParent.apply(item);
-        if (parentNode != null) {
-            builder.addEdge(parentNode, item);
-        }
+        getParent.apply(item)
+                .map(parentNode -> builder.addEdge(parentNode, item));
     }
 
-    public Stream<T> getOrdered() {
+    public Stream<T> getBuildChain() {
         DefaultDirectedGraph<T, DefaultEdge> graph = builder.build();
-        AbstractGraphIterator<T, DefaultEdge> i = new TopologicalOrderIterator<>(graph);
-        Iterable<T> iterable = () -> i;
+        return getStreamFromIterator(new TopologicalOrderIterator<>(graph));
+    }
+
+    private static <T> Stream<T> getStreamFromIterator(Iterator<T> iterator) {
```

Comment:
Can you put this into an (existing?) Util class? It's not the first time we need this.